### DISCUSSION TITLE
Website: should we keep docs of old versions available?

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6056] [Technical Discussions]
Following up on:
* #5939
* #6055 

Right now we only include docs for the latest version of `@typescript-eslint/` packages on the website. That's good enough for most users.

But! It might be nice if we could also include docs for older versions. Sometimes folks are stuck on old versions, or just curious. Docusaurus has a feature for this: https://docusaurus.io/docs/versioning. The feature comes at a cost:

> > **⚠️ Caution**
> > Think about it before starting to version your documentation - it can become difficult for contributors to help improve it!
>
> Most of the time, you don't need versioning as it will just increase your build time, and introduce complexity to your codebase. Versioning is **best suited for websites with high-traffic and rapid changes to documentation between versions**. If your documentation rarely changes, don't add versioning to your documentation.

I do worry about the potential added development cost of onboarding to Docusaurus versioning. Developers being able to contribute to the docs is very important to us. Getting in the way of that is a Bad.

Is there anybody out there who would really want old docs available?

**Comment by bradzacher** [comment]
I personally don't think it's worth-while.
If someone wants old versions then we have GH tags to help them go back in time to specific versions.

**Comment by santiagofn** [comment]
Hi!

In corporate environments it's not unusual to keep major versions unchanged for a long time. That's the case of the project I'm working on nowadays: current typescript-eslint version is 6.5.0, and this project's version is 5.6.0. I needed to find info about version 5.6.0 and failed.

If you ask me, having previous releases documentation available is a must for so widely-used packages (unless you think everyone is using the last version, which is not happening in the corporate world). Using GH tags is not a viable solution for user wanting to read docs, not .md files using GH navigator.  I've read https://docusaurus.io/docs/versioning and the only folder developers should modify if versioning is enabled would be `docs` (assuming they contribute to the new version). But an extra burden will be added to each release (a new `version...` folder should be created, etc.).

**Reply by JoshuaKGoldberg** [reply]
> this project's version is 5.6.0

Out of curiosity, what's holding you back from either:
* Upgrading to 6.5.0? _(and in general staying up-to-date)_
* Having developers check out the repo at `5.6.0` locally? _(I haven't tried this to verify it can even run)_

_(I have guesses such as dev time cost, but want to validate)_

**Reply by bradzacher** [reply]
I'd tend to agree - I can understand not being able to upgrade to v6 because breaking changes can be hard and require dedicated time allocated to work through.

But v5.6.0 isn't even close to the latest v5 release (5.62.0). Your company is _93_ releases behind the latest non-breaking release (which is almost two full years behind).

And it is over 2 full years behind the latest v6 release.

---

Also worth noting that our npm package ships with the rule docs included. So you can also find all documentation for the rules in your version in `node_modules/@typescript-eslint/eslint-plugin/docs/rules`

https://unpkg.com/browse/@typescript-eslint/eslint-plugin@5.6.0/docs/rules/

**Reply by santiagofn** [reply]
Sorry, I confused the version. It's in fact 5.62.0. 

![package.json screenshot](https://github.com/typescript-eslint/typescript-eslint/assets/8375708/149e1211-0538-41f4-8b20-8b9bbf8bbfb7)

As Brad said, when dealing with multiple projects and requests it's not always easy to prioritize bumping versions to a new major.

Thanks for the pkg docs hint. But as I mentioned, it means navigating through .md files. It'd be great to be able to use a webpage.

**Reply by bradzacher** [reply]
The vast majority of docs should be compatible with v5 - rules are pretty static version-to-version and even across the major we don't tend to change much aside from removing things.

**Reply by santiagofn** [reply]
Configurations seem to have changed. I found that v5 config `recommended-requiring-type-checking` has been replaced by `recommended-type-checked` in v6 and that `stylistic...` configs don't exist in v5.

**Reply by cdavie-artium** [reply]
I'm also in a corporate environment, using v5, looking for configuration docs that are different between v5 and v6. I don't necessarily get to choose when we prioritize upgrading old packages vs other work on our plate, so while it's true that we _could_ upgrade, realistically I have no idea when we actually _will_. So I wish it were easier to find the old docs.

**Reply by bradzacher** [reply]
@cdavie-artium you can use the tag switcher on the github UI to select the relevant version tag. You won't get the fully rendered docs in a website - but you'll be able to browse the markdown files which aren't _too_ different.

**Comment by JoshuaKGoldberg** [comment]
Here's how you can spin up a local dev server for any arbitrary tag:

```shell
git clone https://github.com/typescript-eslint/typescript-eslint
cd typescript-eslint
git checkout v5.62.0
yarn
yarn nx run website-eslint:build
yarn start
```

Note that if `yarn start` fails with a message about `export = config`, add this before `yarn start`:

```shell
sed -i '' '/swc: true/d' packages/website/docusaurus.config.js
```

See https://github.com/typescript-eslint/typescript-eslint/issues/7165.

**Reply by santiagofn** [reply]
Thanks, Joshua. I'll look into this.

**Comment by JoshuaKGoldberg** [comment]
Closing this out as it doesn't seem actionable for us. We definitely don't have the time to maintain a deployment with historical versions - even if we use the latest and greatest built-in Docusaurus support. Given that there are several ways to view docs mentioned in this discussion, we don't feel a strong need to take any action. Thanks y'all!
